### PR TITLE
[sk] Calculate the indices of invalid values

### PR DESCRIPTION
**Pull request by skunichetty**
# Summary
Backend now calculates indices of invalid values for phone numbers, zip codes, and emails. This can be used by the frontend to highlight these entries as invalid.

# Tests
Unit tests were developed locally to test that the invalid indices are calculated correctly, alongside testing on user datasets to confirm that the validity metric still reports outputs correctly.

cc:
@wangxiaoyou1993

### CODE REVIEW

File: mage_ai/data_cleaner/column_type_detector.py [hunk 1]
```diff
@@ -63,18 +63,23 @@ def str_in_set(string, string_set):
...
+def find_syntax_errors(series, column_type):
     if len(series) == 0:
-        return []
-    mismatched_rows = []
+        return pd.Series([])
+    mask = pd.Series([False] * len(series))
+    mask.index = series.index
     if column_type == EMAIL:
-        mismatched_rows = series[~series.str.match(REGEX_EMAIL)].tolist()
+        mask |= ~series.str.match(REGEX_EMAIL)
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/column_type_detector.py:72]: why do you need mask? Can you just return `~series.str.match(xxx)`

File: mage_ai/data_cleaner/statistics/calculator.py [hunk 2]
```diff
@@ -231,9 +235,11 @@ def __statistics_overview(self, series, col):
...
 
         # Detect mismatched formats for some column types
-        invalid_rows = get_mismatched_rows(series_non_null, column_type)
-        data[f'{col}/invalid_value_count'] = len(invalid_rows)
-        data[f'{col}/invalid_values'] = invalid_rows[:INVALID_VALUE_SAMPLE_COUNT]
+        invalid_rows = find_syntax_errors(series_non_null, column_type)
+        data[f'{col}/invalid_value_count'] = invalid_rows.sum()
+        invalid_values = series_non_null[invalid_rows]
+        data[f'{col}/invalid_values'] = invalid_values[:INVALID_VALUE_SAMPLE_COUNT].tolist()
+        data[f'{col}/invalid_indices'] = invalid_values.index.tolist()
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/statistics/calculator.py:242]: For invalid indices, we need to return all indices below [SAMPLE_SIZE](https://github.com/mage-ai/mage-ai/blob/master/mage_ai/server/data/models.py#L7).

Comment (skunichetty) [reply mage_ai/data_cleaner/statistics/calculator.py:242]: Ok, is there any way to support sharing that constant between the `server` subpackage and the `data_cleaner` subpackage? If not, I could add a new `shared` subpackage in the `mage_ai` folder from which both `server` and `data_cleaner` could fetch this constant value.

Comment (wangxiaoyou1993) [reply mage_ai/data_cleaner/statistics/calculator.py:242]: just move the constant to data_cleaner subpackage

Comment (skunichetty) [reply mage_ai/data_cleaner/statistics/calculator.py]: Okay I moved this constant into the shared folder in the data_cleaner subpackage.

File: mage_ai/data_cleaner/column_type_detector.py (hunk 1 above)
Comment (skunichetty) [reply mage_ai/data_cleaner/column_type_detector.py]: Fixed

**Comment by skunichetty** [comment]
Note: This latest push also fixed a bug in the case where non-integer values are used as dataframe indices. Now the the integer locations of indexes are returned rather than actual dataframe index itself.

File: mage_ai/data_cleaner/statistics/calculator.py [hunk 3]
```diff
@@ -231,9 +237,12 @@ def __statistics_overview(self, series, col):
...
 
         # Detect mismatched formats for some column types
-        invalid_rows = get_mismatched_rows(series_non_null, column_type)
-        data[f'{col}/invalid_value_count'] = len(invalid_rows)
-        data[f'{col}/invalid_values'] = invalid_rows[:INVALID_VALUE_SAMPLE_COUNT]
+        invalid_rows = find_syntax_errors(series_non_null, column_type)
+        data[f'{col}/invalid_value_count'] = invalid_rows.sum()
+        invalid_values = series_non_null[invalid_rows]
+        data[f'{col}/invalid_values'] = invalid_values[:INVALID_VALUE_SAMPLE_COUNT].tolist()
+        invalid_indices = series.index.get_indexer(invalid_values.index)
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/statistics/calculator.py:244]: is it needed? we didn't reindex the series after filtering null values. we can just use `invalid_values.index`

File: mage_ai/data_cleaner/statistics/calculator.py [hunk 4]
```diff
@@ -231,9 +237,12 @@ def __statistics_overview(self, series, col):
...
+        data[f'{col}/invalid_indices'] = invalid_indices[np.where(invalid_indices <= SAMPLE_SIZE)]
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/statistics/calculator.py:245]: nit: `tolist()`

File: mage_ai/data_cleaner/statistics/calculator.py (hunk 3 above)
Comment (wangxiaoyou1993) [reply mage_ai/data_cleaner/statistics/calculator.py:244]: nvm, just saw your latest comment.

**Review by wangxiaoyou1993 — APPROVED** [review]

**Pull request merged**